### PR TITLE
Add more sections to plugin authors' guide

### PR DESCRIPTION
**Pull request by gedankenexperimenter**
The code snippets are still untested, so I'm marking this as a draft.  I plan to add testcases to verify that the code contained in the guide actually works as described.

### CODE REVIEW

File: docs/customization/plugin-authors-guide.md [hunk 1]
```diff
@@ -250,12 +250,209 @@ kaleidoscope::plugin::MyPlugin;
...
+```c++
+  EventHandlerResult onKeyEvent(KeyEvent &event) {
+    if (event.key == Key_H && keyToggledOn(event.state)) {
+
+	  // Create and send the `H` (shift + h)
+	  KeyEvent new_event = KeyEvent(KeyAddr::none(), IS_PRESSED, LSHIFT(Key_H));
+	  Runtime.handleKeyEvent(new_event);
+
+	  // Change the key value and send `i`
+      new_event.key = Key_I;
```
Comment (The-Compiler) [inline docs/customization/plugin-authors-guide.md:278]: There are some indentation issues, also in the code snippets below.

File: docs/customization/plugin-authors-guide.md [hunk 2]
```diff
@@ -250,12 +250,209 @@ kaleidoscope::plugin::MyPlugin;
...
+
+One very important consideration for any plugin that calls `Runtime.handleKeyEvent()` from an `onKeyEvent()` handler is recursion.  `Runtime.handleKeyEvent()` will call all plugins' `onKeyEvent()` handlers, including the one that generated the event.  Therefore, we need to take some measures to short-circuit the resulting recursive call so that our plugin doesn't cause an infinite loop.
+
+Suppose the example plugin above was changed to type the string `hi!` instead of `Hi!`.  When sending the first generated event, with `event.key` set to `Key_H`, our plugin would recognize that event as one that should be acted on, and make another call to `Runtime.handleKeyEvent()`, which would again call `MyPlugin.onKeyEvent()`, and so on until the MCU ran out of memory on the stack.
+
+The simplest mechanism used by many plugins that mark their generated events "injected" is to simply ignore all generated events:
+
+```c++
+  EventHandlerResult onKeyEvent(KeyEvent &event) {
+    if ((event.state | INJECTED) != 0)
```
Comment (The-Compiler) [inline docs/customization/plugin-authors-guide.md:384]:
```suggestion
    if ((event.state & INJECTED) != 0)
```

surely? There should be a helper for this really I think, so I opened #1082.

**Review by The-Compiler — CHANGES_REQUESTED** [review]
I just found this PR after [having authored my first plugin](https://twitter.com/the_compiler/status/1432025502928285697), and the plugin author's guide is great! This sure would make a good addition. I commented on a couple of things I noticed while reading through it.

File: docs/customization/plugin-authors-guide.md (hunk 1 above)
Comment (gedankenexperimenter) [reply docs/customization/plugin-authors-guide.md:278]: Thanks.  I'll go through it and fix them up.

File: docs/customization/plugin-authors-guide.md (hunk 2 above)
Comment (gedankenexperimenter) [reply docs/customization/plugin-authors-guide.md:384]: Clearly an error, thanks!

File: docs/customization/plugin-authors-guide.md (hunk 1 above)
Comment (gedankenexperimenter) [reply docs/customization/plugin-authors-guide.md:278]: Ah, I see the problem.  I was writing all that in markdown-mode, and tab characters ended up where I didn't expect them.

**Comment by gedankenexperimenter** [comment]
As I'm lacking time and motivation to do more writing on this now, I'll submit it in its current state, since at least one person thinks it's useful (Thanks, @The-Compiler).

**Comment by obra** [comment]
And merged :)

**Pull request merged**